### PR TITLE
Cleanup TF todo items

### PR DESCRIPTION
**Pull request by Aaron-Robertson**
**Context:**
Items marked `todo` during initial implementation of the `tfbackend` are ready for cleanup.

**Description of the Change:**
Resolves cleanup items in `tfbackend` based on updated version of `tensorflow`.

**Benefits:**
Resolves `todo` items to reduce line count, clarify function similarity across backend ops, and clean the approach to `tensorflow.tensordot` based on TF version >2.0.

**Possible Drawbacks:**
N/A

**Related GitHub Issues:**
N/A

### CODE REVIEW

**Comment by co9olguy** [comment]
Hi @Aaron-Robertson, just wanted to check the status of this PR. Any questions the team could help answer?

**Comment by Aaron-Robertson** [comment]
Hey @co9olguy, this one covers the low hanging cleanup items so far. I wanted improve partial trace but haven't gotten a chance to try much beyond research so far. The rest of the todo marked items in the `tfbackend` are tied to that partial trace, so if there's a recommendation for path forward I plan to take a shot at it this weekend.

I figure I could have more specific questions next week once I try some things.

**Comment by co9olguy** [comment]
No worries @Aaron-Robertson :slightly_smiling_face: 

We're also happy to accept piecewise PRs (not all cleanup items need to be in one PR), so long as the updates are independent

**Comment by Aaron-Robertson** [comment]
Fair enough, I'll get this one ready to merge then and can start a separate one for the heavier lift 😄

File: strawberryfields/backends/tfbackend/ops.py [hunk 1]
```diff
@@ -70,7 +70,7 @@ def _numer_safe_power(base, exponent, dtype=tf.complex64):
     return base ** exponent
 
 
-def mixed(pure_state, batched=False):
+def mix(pure_state, batched=False):
     """Converts the state from pure to mixed"""
     # todo: In the fock backend mixing is done by ops.mix(), maybe the functions should be named identically?
```
Comment (josh146) [inline strawberryfields/backends/tfbackend/ops.py:75]: Looks like you can delete this TODO comment now, @Aaron-Robertson :)

File: strawberryfields/backends/tfbackend/ops.py [hunk 2]
```diff
@@ -931,24 +931,11 @@ def replace_modes(replacement, modes, system, system_is_pure, batched=False):
         # append mode and insert state (There is also insert_state(), but it seemed unnecessarily complicated to try to generalize this function, which does a lot of manual list comprehension, to the multi mode case than to write the two liner below)
         # todo: insert_state() could be rewritten to take full advantage of the high level functions of tf. Before doing that different implementatinos should be benchmarked to compare speed and memory requirements, as in practice these methods will be perfomance critical.
         if not batched:
-            # todo: remove the hack in the line below and enable the line with axes=0 instead, if ever we change the dependency of SF to tensorflow>=1.6
-            # revised_modes = tf.tensordot(reduced_state, replacement, axes=0)
-            revised_modes = tf.tensordot(
-                tf.expand_dims(reduced_state, 0), tf.expand_dims(replacement, 0), axes=[[0], [0]]
-            )
+            revised_modes = tf.tensordot(reduced_state, replacement, axes=0)
```
Comment (josh146) [inline strawberryfields/backends/tfbackend/ops.py:934]: Nice! This should also help with speed I imagine?

I'm just tagging @ziofil for any thoughts or comments.

File: .github/CHANGELOG.md [hunk 3]
```diff
@@ -98,6 +98,9 @@
   method.
   [(#564)](https://github.com/XanaduAI/strawberryfields/pull/564)
 
+* Cleanup `backends/tfbackend/ops.py` todo items.
```
Comment (josh146) [inline .github/CHANGELOG.md:101]: I would make this more specific - the changelog is intended to give users and developers a quick overview of any changes since the last release, without them needing to look at the code :)

So better to instead list the specific changes/improvements made.

**Review by josh146 — CHANGES_REQUESTED** [review]
Thanks @Aaron-Robertson, good to get these todo's addressed!

It looks good to me, although would be happy to get @ziofil's thoughts on the TF changes. Other than that, it's just a matter of being more specific in the changelog entry :)

**Review by thisac — APPROVED** [review]
Looks great @Aaron-Robertson! Apart from the comments from @josh146, I think it looks good. Thanks for working on this! 👍

File: strawberryfields/backends/tfbackend/ops.py (hunk 2 above)
Comment (ziofil) [reply strawberryfields/backends/tfbackend/ops.py:934]: I made a few experiments and I found that broadcasting is about 40x faster than the method above. it can be done programmatically by writing the indices like so:
```python
num_modes = len(reduced_state.shape) - 1
idx1 = (slice(None,None),)+(Ellipsis,) + (None,)* num_modes
idx2 = (slice(None,None),)+(None,)* num_modes +(Ellipsis,)
revised_modes = reduced_state[idx1] * replacement[idx2]
```

Comment (ziofil) [reply strawberryfields/backends/tfbackend/ops.py:934]: one could even bypass the `if batched / else` block with 
```python
idx1 = (slice(None,None),)*batched + (Ellipsis,) + (None,)*(replacement.ndim-int(batched))
idx2 = (slice(None,None),)*batched + (None,)*(reduced_state.ndim-int(batched)) + (Ellipsis,)
revised_modes = reduced_state[idx1] * replacement[idx2]
```

Comment (josh146) [reply strawberryfields/backends/tfbackend/ops.py:934]: Oh, that's awesome. Is this generally the case that broadcasting is faster than tensordot? Or just for this particular application?

Comment (Aaron-Robertson) [reply strawberryfields/backends/tfbackend/ops.py]: Made and pushed these changes. Certainly a speedup in general, though specific cases are about equivalent (e.g. homodyne).

File: strawberryfields/backends/tfbackend/ops.py (hunk 1 above)
Comment (ziofil) [reply strawberryfields/backends/tfbackend/ops.py]: > Oh, that's awesome. Is this generally the case that broadcasting is faster than tensordot? Or just for this particular application?

Here we used it for a batched/nonbatched outer product, but it seems to be the case for a few applications, yes:
```python
T1 = tf.constant(np.random.normal(size=(13,5,11,17)))
T2 = tf.constant(np.random.normal(size=(13,5,11,17)))
```
Tensordot runtime:
```python
%timeit tf.tensordot(T1,T2,[[0,1,2,3],[0,1,2,3]])
# 127 µs ± 950 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```
broadcasting is about 5x faster
```python
%timeit tf.reduce_sum(T1*T2)
# 25.2 µs ± 989 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```
but `einsum` seems even better:
```python
%timeit tf.einsum('ijkl,ijkl', T1, T2)
# 20.9 µs ± 791 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```
The only advantage at this point is that the broadcasting approach may make the code simpler with respect to `einsum` (no convoluted strings of indices to generate separately).

**Comment by Aaron-Robertson** [comment]
@josh146 I think this one's ready to go. Merged master down with latest changes and it looks like the latest `black` version is throwing an error on comment format where it wasn't before. Can fix here on request.

File: .github/CHANGELOG.md [hunk 4]
```diff
@@ -137,6 +137,11 @@ This release contains contributions from (in alphabetical order):
   method.
   [(#564)](https://github.com/XanaduAI/strawberryfields/pull/564)
 
+* Cleanup `backends/tfbackend/ops.py` to reduce line count, clarify function
+  similarity across backend ops, and replace `tensorflow.tensordot` with
+  broadcasting.
+  [(#567)](https://github.com/XanaduAI/strawberryfields/pull/567)
```
Comment (josh146) [inline .github/CHANGELOG.md:143]: We've recently released 0.18, so this can be moved up to the v0.19 section of the changelog :) Don't forget to add your name to the contribution section there as well!

**Review by josh146 — APPROVED** [review]
Looks great @Aaron-Robertson! Thanks for this.

Re: black, I suggest upgrading the local version of black that you have, and re-blacking the codebase, in order to get the check passing.

File: .github/CHANGELOG.md (hunk 4 above)
Comment (Aaron-Robertson) [reply .github/CHANGELOG.md]: Ready to go! Apologies for the delay 😄
